include/cetl/span.h

```diff
@@ -1 +1,772 @@
+    }
+
+    /// Create a new span from the start of the current span for count elements.
+    /// @param count  The number of elements for the sub-span. Where `count` is greater than the
+    ///               current span's size the behavior is undefined.
+    /// @return A new span with a dynamic extent extent and a size of `count`.
+    /// @see std::span::first
+    constexpr span<element_type, cetl::dynamic_extent> first(size_type count) const
+    {
+        CETL_DEBUG_ASSERT(count <= Extent, "Dynamic subviews beyond the size of the span's view are undefined.");
+        return span<element_type, cetl::dynamic_extent>(data_, count);
+    }
+
+    /// Create a new span `Count` elements from the last item of the current span to its end.
+    /// @tparam Count   Number of elements for the sub-span.
+    /// @return A new span with the static extent `Count`.
+    /// @see std::span::last
+    template <size_type Count>
+    constexpr span<element_type, Count> last() const
+    {
+        static_assert(Count <= extent, "subviews beyond the size of the span's view are not allowed.");
+        return span<element_type, Count>{(data_ + extent) - Count, Count};
+    }
+
+    /// Create a new span `count` elements from the last item of the current span to its end.
+    /// @param count   Number of elements for the sub-span. The behavior of this method is undefined where
+    ///                `count` is greater than the current span's size.
+    /// @return A new span with a size of `count` and a dynamic extent.
+    /// @see std::span::last
+    constexpr span<element_type, cetl::dynamic_extent> last(size_type count) const
+    {
+        CETL_DEBUG_ASSERT(count <= extent, "Dynamic subviews beyond the size of the span's view are undefined.");
+        return span<element_type, cetl::dynamic_extent>((data_ + extent) - count, count);
+    }
+
+    /// Create a new span `Offset` elements from the start of the current span and for `Count` elements.
+    /// @tparam Offset  Number of elements from the start of the current span for the subspan.
+    /// @tparam Count   The number of elements from the Offset to include in the subspan. If this value is
+    ///                 `cetl::dynamic_extent` then the Count is `size() - Offset`.
+    /// @return A new span where the `Extent` is `Count` if this was not `cetl::dynamic_extent` otherwise the
+    ///         new span's `Extent` is this span's `Extent` minus `Offset`.
+    /// @see std::span::subspan()
+    template <size_type Offset, size_type Count = cetl::dynamic_extent>
+    constexpr auto subspan() const -> span<element_type, (Count != cetl::dynamic_extent) ? Count : Extent - Offset>
+    {
+        static_assert(Offset <= extent, "subspan Offsets > extent are ill-formed.");
+        static_assert((Count == cetl::dynamic_extent) || (Count <= extent - Offset),
+                      "subspan Count argument is ill-formed");
+        return span < element_type,
```

Comment:
this looks like a comma operator with a comparisons.  Not sure if there's a better way to write it. maybe just a comment?

Reply:
The way the code is formatted makes it hard to follow. Was it clang-format? If so, a newer version may fix this as there've been many issues of this kind before LLVM 15.

Reply:
it is clang-format from the ts20.4.1 container. We'd need to update it there.

Reply:
I was using the wrong clang-format. I just pushed a change that is formatted using the Toolshed container version. Let me know if this is what you expected.

Reply:
It's still the same. Maybe it's time to update the container? On the other hand, this is just formatting, it doesn't affect validity so this issue is a very low-priority one. We can upgrade the container later.

Reply:
https://github.com/OpenCyphal/docker_toolchains/issues/25

Reply:
I just mean that the  `span < element_type`  looks like testing if `span` is `<` than `element_type`

Reply:
That's what old clang-format does to templates, yes.